libkbfs/file_data.go

```diff
@@ -97,6 +122,341 @@ func (fd *fileData) getFileBlockAtOffset(ctx context.Context,
 	return ptr, parentBlocks, block, nextBlockStartOff, startOff, wasDirty, nil
 }
 
+// getLeafBlocksForOffsetRange fetches all the leaf ("direct") blocks
+// that encompass the given offset range (half-inclusive) in the file.
+// If `endOff` is -1, it returns blocks until reaching the end of the
+// file.  Note the range could be made up of holes, indicated in the
+// IPtrs list of the parent blocks.  If `prefixOk` is true, the
+// function will ignore context deadline errors and return whatever
+// prefix of the data it could fetch within the deadine.  Return
+// params:
+//
+//   * pathsFromRoot is a slice, ordered by file offset, of paths from
+//     the root to each block that makes up the range.
+//   * blocks: a map from block pointer to a data-containing leaf node
+//     in the given range of offsets.
+//   * nextBlockOff is the offset of the block that follows the last
+//     block given in `pathsFromRoot`.  If `pathsFromRoot` contains
+//     the largest block among the children, nextBlockOff is -1.
+func (fd *fileData) getLeafBlocksForOffsetRange(ctx context.Context,
+	pblock *FileBlock, startOff int64, endOff int64,
+	prefixOk bool) (pathsFromRoot [][]parentBlockAndChildIndex,
+	blocks map[BlockPointer]*FileBlock, nextBlockOffset int64, err error) {
+	if !pblock.IsInd {
+		// Return a single empty path and am empty child block map.
+		return [][]parentBlockAndChildIndex{nil}, nil, -1, nil
+	}
+
+	type resp struct {
+		pathsFromRoot [][]parentBlockAndChildIndex
+		blocks        map[BlockPointer]*FileBlock
+		nextBlockOff  int64
+	}
+
+	// Search all of the in-range child blocks, and their child
+	// blocks, etc, in parallel.
+	respChans := make([]<-chan resp, 0, len(pblock.IPtrs))
+	eg, groupCtx := errgroup.WithContext(ctx)
+	nextBlockOff := int64(-1)
+	for i, iptr := range pblock.IPtrs {
+		// Some byte of this block is included in the left side of the
+		// range if `startOff` is less than the largest byte offset in
+		// the block.
+		inRangeLeft := true
+		if i < len(pblock.IPtrs)-1 {
+			inRangeLeft = startOff < pblock.IPtrs[i+1].Off
+		}
+		if !inRangeLeft {
+			continue
+		}
+		// Some byte of this block is included in the right side of
+		// the range if `endOff` is bigger than the smallest byte
+		// offset in the block (or if we're explicitly reading all the
+		// data to the end).
+		inRangeRight := endOff == -1 || endOff > iptr.Off
+		if !inRangeRight {
+			// This block is the first one past the offset range
+			// amount the children.
+			nextBlockOff = iptr.Off
+			break
+		}
+
+		ptr := iptr.BlockPointer
+		childIndex := i
+		respCh := make(chan resp, 1)
+		respChans = append(respChans, respCh)
+		eg.Go(func() error {
+			block, err := fd.goGetter(groupCtx, fd.kmd, ptr, fd.file)
+			if err != nil {
+				return err
+			}
+			// Recurse down to the level of the child.
+			pfr, blocks, nextBlockOff, err := fd.getLeafBlocksForOffsetRange(
```

Comment:
Right now because you're making individual goroutines for all `IPtrs` retrievals, they can race against each other, at all levels, so we have no way to ensure they get downloaded in a sane order. It might work better if we create a `BlockOps.GetAsync` that is called synchronously, where this `eg.Go` goroutine is spawned after we get back the channel from that call.